
<div style="border-radius:10px;
            border : #015a2c solid;
            background-color:#D7D9DB;
           font-size:110%;
           letter-spacing:0.5px;
            text-align: center">
​
<center><h1 style="padding: 25px 0px; color:#015a4c; font-weight: bold; font-family: Cursive">
Amit Project for Data Science Module <br><br>💡💡💡💡💡 <br><br> DEPI_BS_DS_R3  </h1></center>
<center><h3 style="padding-bottom: 25px; color:#025b2c; font-weight: bold; font-style:italic; font-family: Cursive">
Using Project: Trip data Ford go bike  </h3></center>     
​
</div>

# Part I - (Dataset Exploration for Ford GoBike)

### worked by (Manwel Wasfy Georgy)

## Introduction
======================================

.

**This data set information was produced from `Ford GoBike` System** the information in it talked about individual rides that were made in a bike-sharing system for covering San Francisco Bay area.** 

.




> ## And here are the main exploring questions 



   **==================================================**



- What is the structure of your dataset?


- What are the main interesting features in the dataset?


- What features in the dataset supported my investigation for these interesting features?




.

> ### Also here are the investigative questions classified into three types



   **================================================**
   
   


###  <a href='#011'> Univariate Exploration</a> 

- <a href='#1'> 1 ) . What is the distribution of `User type`?</a>  
- <a href='#2'> 2 ) . What is the distribution for the `bike share users on all Trips`?</a>  
- <a href='#3'> 3 ) . What is the distribution for users' `ages`?</a>  
- <a href='#4'> 4 ) . Do the users' ages have `outlier` values?</a>
- <a href='#5'> 5 ) . What is the distribution for the trip duration in `minutes`? </a>
- <a href='#6'> 6) . What is the distribution for `member_gender`? </a>
- <a href='#7'> 7) . What is the distribution for the trip duration in `Hours`?  </a>




##  <a href='#111'> Bivariate Exploration</a> 


- <a href='#8'> 8 ) . What the `average duration` of trip for the categories of `user type`? </a>
- <a href='#9'> 9 ) . How long is the duration of the trip for `each gender` in `minutes`? </a>
- <a href='#10'> 10 ) . How long is the duration of the trip for each `user_type` in `minutes`?  </a>
- <a href='#11'> 11 ) . How the user distribution depending on their `Age` and `user_type` ?</a>
- <a href='#12'>12 ). How the user distribution depending on their `Age` and `member_gender` ?  </a>


##  <a href='#1111'> Multivariate Exploration</a> 


- <a href='#13'> 13 ) . What is the Correlation between `each` variable in the dataset?</a>  
- <a href='#14'> 14 ) . What is the Correlation between `specific numeric variables` in the dataset?</a>  
- <a href='#15'> 15 ) . What is the  Correlation between `data_stat` variables in the dataset?</a>  


##  <a href='#1100'> Conclusions </a> 





# Dataset metadata
Scope: Covers only a single month (February 2019). Do not run cross-month seasonality or annual trend analysis.


# Dataset decompisition
**Temporal** = [duration_sec, start_time, end_time]

**ids&names** = [start_station_id, start_station_name, end_station_id, end_station_name, bike_id] 
    since we are not training a ml model we don't need to drop and could extract facts

**Spatial** = [start_station_latitude, start_station_longitude, end_station_latitude, end_station_longitude]

**User informations** = [user_type, member_birth_year, member_gender]

**Econmic indicator** = [bike_share_for_all_trip]

*nulls exists

    duration_sec => compared to (user_type, birth_year, gender, bike_share_for_all_trip)


    start_time, end_time => already aggregated in duration_sec and doesn't include the hour but could answer smth like 
at which part of the hour do most users start renting, Separate circular trips

    start_station_id, start_station_name, end_station_id, end_station_name => stations with the most rides orignating from, 
stations with the most rides ending at, connected stations

    start_station_latitude, start_station_longitude, end_station_latitude, end_station_longitude => Haversine displacement
distance, Riding speed,

    bike_id => most used bikes, bikes usage compared to (duration, user_type, birth_year, gender)

    user_type, member_gender, bike_share_for_all_trip => category columns with lots of context, should ne used heavily

    bike_share_for_all_trip => Analyze whether low-income program participants use bikes differently

### Dataset Decomposition & Analysis Plan

**Scope:** February 2019 (Cross-month seasonality and annual trend analysis are out of scope).

---

| Category | Columns | Planned Engineering & Analytical Goals |
| :--- | :--- | :--- |
| **Temporal** | `duration_sec`<br>`start_time`<br>`end_time` | • Convert duration to minutes; apply log scaling to handle extreme skews.<br>• Bin start minutes (`0–10`, `10–20`, etc.) to evaluate transit connection patterns within the hour.<br>• Cross-tabulate duration against demographics, user type, and equity program enrollment. |
| **Network & Stations** | `start_station_id`<br>`start_station_name`<br>`end_station_id`<br>`end_station_name` | • Identify top origin, destination, and paired travel corridors.<br>• Calculate net station flow ($\text{Arrivals} - \text{Departures}$) to pinpoint bike surplus and deficit hubs.<br>• Flag circular round-trips (`start_station_id == end_station_id`) vs. one-way transit. |
| **Spatial** | `start_station_latitude`<br>`start_station_longitude`<br>`end_station_latitude`<br>`end_station_longitude` | • Calculate straight-line displacement via the Haversine formula.<br>• Approximate speed ($\text{Distance} / \text{Duration}$) to detect non-movement outliers or leisure cruising.<br>• Map geographic trip clusters across San Francisco, Oakland, and San Jose. |
| **Demographics** | `user_type`<br>`member_birth_year`<br>`member_gender` | • Derive user age (`2019 - member_birth_year`) and group into demographic cohorts (e.g., `<25`, `25–34`, `35–54`, `55+`).<br>• Filter implausible birth years (e.g., ages $>80$).<br>• Compare ride duration across gender identities and user subscription models (`Subscriber` vs. `Customer`). |
| **Equity Program** | `bike_share_for_all_trip` | • Measure equity program adoption among eligible riders.<br>• Evaluate whether low-income program participants differ in trip lengths, repeat usage, or primary station hubs. |
| **Fleet Tracking** | `bike_id` | • Rank bikes by cumulative usage (ride counts and total operating hours).<br>• Identify heavily utilized inventory for proactive maintenance flags. |

---

### Data Hygiene & Null Strategy

* **Station Metadata (`start_station_id`, `end_station_id`):** 197 missing records (~0.1%). Exclude from station-level routing analyses or tag as dockless/off-grid trips; keep for aggregate duration metrics.
* **Demographics (`member_birth_year`, `member_gender`):** 8,265 missing records (~4.5%), concentrated in casual `Customer` trips. Segment demographic analyses exclusively to authenticated `Subscriber` data rather than imputing artificial values.

# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Import Libraries</p>

In [ ]:
import numpy as np
import pandas as pd
import os
from pathlib import Path
import matplotlib.pyplot as plt
import seaborn as sns


In [ ]:
# Load dataset (CSV lives next to this notebook in Project/Solo)
csv_path = Path("fordgobike-tripdataFor201902.csv")
if not csv_path.exists():
    csv_path = Path("/home/manwelw/depi/Depi/data-analysis/Project/Solo/fordgobike-tripdataFor201902.csv")
df = pd.read_csv(csv_path)
df.head()


In [59]:
# here we make overview in high-level for data shape and composition
df.shape

(183416, 16)

In [60]:
## check each var  datatype 
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 183416 entries, 0 to 183415
Data columns (total 16 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   duration_sec             183416 non-null  int64  
 1   start_time               183416 non-null  str    
 2   end_time                 183416 non-null  str    
 3   start_station_id         183219 non-null  float64
 4   start_station_name       183219 non-null  str    
 5   start_station_latitude   183416 non-null  float64
 6   start_station_longitude  183416 non-null  float64
 7   end_station_id           183219 non-null  float64
 8   end_station_name         183219 non-null  str    
 9   end_station_latitude     183416 non-null  float64
 10  end_station_longitude    183416 non-null  float64
 11  bike_id                  183416 non-null  int64  
 12  user_type                183416 non-null  str    
 13  member_birth_year        175151 non-null  float64
 14  member_gender  

In [61]:
# descriptive statistics for numeric variables
df.describe().T

,count,mean,std,min,25%,50%,75%,max
duration_sec,183416.0,726.067311,1794.371831,61.000000,325.000000,514.000000,796.000000,85444.000000
start_station_id,183219.0,138.590365,111.778938,3.000000,47.000000,104.000000,239.000000,398.000000
start_station_latitude,183416.0,37.771221,0.099586,37.317298,37.770083,37.780760,37.797280,37.880222
start_station_longitude,183416.0,-122.352661,0.117100,-122.453705,-122.412408,-122.398285,-122.286533,-121.874119
end_station_id,183219.0,136.249401,111.514916,3.000000,44.000000,100.000000,235.000000,398.000000
end_station_latitude,183416.0,37.771425,0.099494,37.317298,37.770407,37.781010,37.797320,37.880222
end_station_longitude,183416.0,-122.352247,0.116677,-122.453705,-122.411726,-122.398279,-122.288045,-121.874119
bike_id,183416.0,4472.890260,1664.386068,11.000000,3777.000000,4958.000000,5502.000000,6645.000000
member_birth_year,175151.0,1984.806504,10.116589,1878.000000,1980.000000,1987.000000,1992.000000,2001.000000


In [62]:
df.dtypes

duration_sec                 int64
start_time                     str
end_time                       str
start_station_id           float64
start_station_name             str
start_station_latitude     float64
start_station_longitude    float64
end_station_id             float64
end_station_name               str
end_station_latitude       float64
end_station_longitude      float64
bike_id                      int64
user_type                      str
member_birth_year          float64
member_gender                  str
bike_share_for_all_trip        str
dtype: object



> ### What is the structure of your dataset?





- The structure of the dataset contains 16 columns and 183416 rows.
 
 
- Most variables are categorical in nature , and here are **dtypes summary :**

        duration_sec                 int64
        start_time                     str
        end_time                       str
        start_station_id           float64
        start_station_name             str
        start_station_latitude     float64
        start_station_longitude    float64
        end_station_id             float64
        end_station_name               str
        end_station_latitude       float64
        end_station_longitude      float64
        bike_id                      int64
        user_type                      str
        member_birth_year          float64
        member_gender                  str
        bike_share_for_all_trip        str
        dtype: object

> ### What are the main interesting features in the dataset?




 **We got three main feature areas which are :**
 
 
 
 
- Temporal
 
 
 
- Demographics
 
  
 
  
- Network & Stations




> ### What features in the dataset supported my investigation for these interesting features?



- start_station_id, start_station_name

- end_station_id, end_station_name

- bike_id

- user_type (Customer, Subscriber)

- member_gender (Male, Female, Other)

- bike_share_for_all_trip (Yes, No)

## <mark>The above result show some missing values exist in some columns So let's display them </mark>

In [ ]:
df.isna().sum()
# NOTE: member_birth_year/member_gender nulls are NOT only Customers.
# Cell 21 check shows ~3.2k Customers AND ~5k Subscribers missing demographics,
# so do not attribute all missing demographics to casual rides.
# Station nulls (197) will be dropped; demographic nulls will be retained (see below).




> ## Missing-data strategy

- **Station columns** (`start_station_id/name`, `end_station_id/name`, 197 rows, ~0.1%): drop via `dropna(subset=...)`. Too few to impute; needed for routing analysis.

- **Demographics** (`member_birth_year`, `member_gender`, 8,265 rows, ~4.5%): **do NOT impute** with mean/mode — that would fabricate ages/genders. Keep as NaN and segment demographic analyses to non-null rows.

- **Duplicates** (4 rows): inspect, then drop (see duplicate cell below).


In [ ]:
# Drop the 197 rows missing station id/name (needed for origin/destination analysis).
# This is a drop, not a fill: imputing station names would fabricate routes.
cols_to_drop = ["start_station_id", "start_station_name", "end_station_id", "end_station_name"]
df = df.dropna(subset=cols_to_drop)


In [65]:
#check the Nulls 
df.isna().sum()

duration_sec                  0
start_time                    0
end_time                      0
start_station_id              0
start_station_name            0
start_station_latitude        0
start_station_longitude       0
end_station_id                0
end_station_name              0
end_station_latitude          0
end_station_longitude         0
bike_id                       0
user_type                     0
member_birth_year          8263
member_gender              8263
bike_share_for_all_trip       0
dtype: int64

### **And with a proper category for other (member gender)**

In [66]:
df['member_gender'].value_counts()

member_gender
Male      130504
Female     40805
Other       3647
Name: count, dtype: int64

In [67]:
df['user_type'].value_counts()

user_type
Subscriber    163418
Customer       19801
Name: count, dtype: int64

In [68]:
##  still got a missing value in the member_gender, so let's explore this column more 
df[ (df['user_type'] == 'Customer') & (df['member_gender'].isna())].count()

duration_sec               3235
start_time                 3235
end_time                   3235
start_station_id           3235
start_station_name         3235
start_station_latitude     3235
start_station_longitude    3235
end_station_id             3235
end_station_name           3235
end_station_latitude       3235
end_station_longitude      3235
bike_id                    3235
user_type                  3235
member_birth_year             0
member_gender                 0
bike_share_for_all_trip    3235
dtype: int64

In [ ]:
# Demographic nulls: quantify by user_type instead of filling with fake values.
print("Remaining birth_year nulls:", int(df["member_birth_year"].isna().sum()))
print("Remaining gender nulls:", int(df["member_gender"].isna().sum()))
print(df.groupby("user_type", observed=True)[["member_birth_year", "member_gender"]].apply(lambda g: g.isna().sum()))
# Strategy: keep NaN; filter to non-null only inside age/gender plots (e.g. df.dropna(subset=['member_age']))


In [ ]:
# Sanity check: Customers missing gender vs Subscribers missing gender
print("Customers missing gender:", int(((df['user_type'] == 'Customer') & (df['member_gender'].isna())).sum()))
print("Subscribers missing gender:", int(((df['user_type'] == 'Subscriber') & (df['member_gender'].isna())).sum()))


In [ ]:
## Result: ~3.2k Customers + ~5k Subscribers lack demographics, so missingness is not 'just casual rides'.
print("Total rows after station drop:", len(df))


> ## Station nulls are resolved (197 rows dropped). Demographic nulls (~8.3k) are intentionally retained as NaN — see strategy above.

### Now let's check duplicate value

In [ ]:
## Check duplicates, inspect, then drop exact duplicate rows
print("Duplicate rows:", int(df.duplicated().sum()))
# display(df[df.duplicated(keep=False)])  # uncomment to inspect the 2 pairs
df = df.drop_duplicates().reset_index(drop=True)
print("Rows after dedup:", len(df))


> ## Duplicates (4 rows / 2 pairs) dropped via `drop_duplicates()` and index reset.


### Now let's check the category columns to make more cleaning dataset 

In [73]:
### check the category columns that included objects
cat_cols = ['user_type', 'member_gender', 'bike_share_for_all_trip']

for col in cat_cols:
    print(f"--- Column: {col} ---")
    print(f"Unique values: {df[col].unique()}")
    print(df[col].value_counts(dropna=False))
    print()

--- Column: user_type ---
Unique values: <StringArray>
['Customer', 'Subscriber']
Length: 2, dtype: str
user_type
Subscriber    163418
Customer       19801
Name: count, dtype: int64

--- Column: member_gender ---
Unique values: <StringArray>
['Male', nan, 'Other', 'Female']
Length: 4, dtype: str
member_gender
Male      130504
Female     40805
NaN         8263
Other       3647
Name: count, dtype: int64

--- Column: bike_share_for_all_trip ---
Unique values: <StringArray>
['No', 'Yes']
Length: 2, dtype: str
bike_share_for_all_trip
No     165872
Yes     17347
Name: count, dtype: int64



 > ## From all previous results above I found  :
 
 
 
 **=========================================================**
 
 

> 1 ) **unique Value** as in ((`member_gender`)) : Not significant , so won't clean it



> 2 ) **Incorrect data type** as in (( `user_type` &&	`member_gender`)) :  significant change , so I need to clean it.


> 3 ) **Good factor** for marketing work as in ((`member_birth_year`)) : so I need to clean it.

In [74]:
##  step  2 ) Incorrect data type as in (( `user_type` &&	`member_gender`)) :  significant change , so  need to clean it.
df['user_type'] = df['user_type'].astype('category')
df['member_gender'] = df['member_gender'].astype('category')
df['bike_share_for_all_trip'] = df['bike_share_for_all_trip'].astype('category')

In [75]:
## check your work 
df[cat_cols].dtypes


user_type                  category
member_gender              category
bike_share_for_all_trip    category
dtype: object

In [ ]:
## placeholder


In [77]:
## check your work 
df['member_age'] = df['member_age'].where((df['member_age'] >= 16) & (df['member_age'] <= 95), np.nan)
df.member_age.isna().sum()


np.int64(8338)

In [78]:
## change the new column age data type from float to int



In [79]:
## create a function to know the average users ages


In [80]:
## another important factor to offer  marketing services is the trip duration time 

## need to change the time from second to minute to be more usable




In [81]:
## I need to change the time from second to Houre  to be more usable



In [82]:
## check your work 



# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Univariate Exploration</p>
# <a id="011"> Univariate Exploration</a>



> In this section, investigate distributions of individual variables. If
you see unusual points or outliers, take a deeper look to clean things up
and prepare yourself to look at relationships between variables.




# <a id="1"> 1 ) . What is the distribution for User type?</a>

In [83]:
## I will import seaborn library to support me in making visualizations 

import seaborn as sns
# plt.figure(figsize = (10,5), dpi = 200)

# plot = sns.countplot(data=gbf, x='bike_share_for_all_trip')

# plt.xlabel('User distribution')
# plt.ylabel('Frequency')
# plt.title("Represent the user's distribution depending on being a part of the Trips Program")

# plt.show()

> ### <mark>Your analsys and exploration here </mark>

# <a id="2"> 2 ) What is the distribution for the bike share users on all Trips?</a>

> ### <mark>Your analsys and exploration here </mark>

# <a id="3"> 3 ) What is the distribution for users' ages? </a>

> ### <mark>Your analsys and exploration here </mark>


# <a id="4">  4 ) . Do the users' ages have outlier values?</a>


> ### <mark>Your analsys and exploration here </mark>

# <a id="5">  5 ) . What is the distribution for the trip duration in minutes? </a>

In [84]:
## Review the statistic for trip duration in minutes



In [85]:
# Make visualization to easy get insight


> ### <mark>Your analsys and exploration here </mark>


# <a id="6"> 6 ) . What is the distribution for member_gender?</a>

In [86]:
# Make visualization to easy get insight


> ### <mark>Your analsys and exploration here </mark>

# <a id="7">    7 )  . What is the distribution for the trip duration in Hours?  </a>

In [87]:
# Make visualization to easy get insight


> ### <mark>Your analsys and exploration here </mark>

### The variables distribution:


> The distribution of duration present that most of the trips were less than 30 minutes ( we can notice that from the minutes and the hour charts )


> And the Age distribution is ( left Skewed ) as most users age are  22 - 39


> Most users are subscribers, not Customers and  Males as shown in the charts above.



### Were there any interesting and outliers 


- We got Age outliers for for more than 100 .



- The distribution of duration was strange where the minutes and the hour charts were 30 minutes or less.

# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Bivariate Exploration</p>
# <a id="111"> Bivariate Exploration</a>

Now let's investigate relationships between pairs of variables in our data.

# <a id="8"> 8) . What the average duration of trip for the categories of user type?</a>

In [88]:
# Distribution of Age by user types



# Make visualization to easy get insight


> ### <mark>Your analsys and exploration here </mark>

# <a id="9">    9 ) . How long is the duration of the trip for each gender in minutes?  </a>

In [89]:
# Make visualization to easy get insight

# ------------------------Duration distribution by member gender



# Make visualization to easy get insight


> ### <mark>Your analsys and exploration here </mark>

# <a id="10 "> 10 ) . How long is the duration of the trip for each user_type in minutes?  </a>

In [90]:

# Make visualization to easy get insight


> ### <mark>Your analsys and exploration here </mark>

# <a id="11">  11 ) . How the user distribution depending on their Age and user_type ? </a>

In [91]:

# Make visualization to easy get insight


> ### <mark>Your analsys and exploration here </mark>

#  <a id="12"> 12 ). How the user distribution depending on their Age and member_gender ?</a>

In [92]:
# Make visualization to easy get insight


> ### <mark>Your analsys and exploration here </mark>

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

> ...................


### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

> ....................

# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Multivariate Exploration</p>
# <a id="1111"> Multivariate Exploration </a>

> Create plots of three or more variables to investigate your data even
further. Make sure that your investigations are justified, and follow from
your work in the previous sections.

In [93]:
## check your work 



# <a id="13">  13 ) . What is the  Correlation between each variable in the dataset?</a>

In [94]:
# Make visualization to easy get insight




> ### <mark>Your analsys and exploration here </mark>

Example usage
- There is `no Correlation` for any variable with time ( `duration sec /minutes/hours`)



# <a id="14"> 14 )  .  What is the  Correlation between `specific numeric variables` in the dataset?</a>


In [95]:
# Make visualization to easy get insight




> ### <mark>Your analsys and exploration here </mark>

Example usage
- There is `no Correlation` for any variable with time ( `duration sec /minutes`)




- 




- 




# <a id="15">  15 ) What is the  Correlation between data_stat variables in the dataset?</a>


In [96]:

"""""This time I will use heatmap  to explain how Correlation between data_stat variables

in the dataset in more clear and specific percentage"""

# Make visualization to easy get insight



'""This time I will use heatmap  to explain how Correlation between data_stat variables\n\nin the dataset in more clear and specific percentage'

> ### <mark>The above charts show how Correlation between specific numeric variables in the dataset using heatmap plots  </mark>



**Actually the `heat map`  display the above ideas in clear percentages**



- There is `no Correlation` for any variable with time ( `duration sec /minutes`)




- There is a `negative  Correlation` between  member_birth_year  and age which normal.




- There is `no significant Correlation` for any variable with `bike_id `



# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Conclusions</p>
# <a id="1100"> Conclusions </a>

#example
- Most users were Male.


- 

- 

- 



## limitation 
example
- We got missing data which we fix some of them .

- 
